fvm/state/accounts.go

```diff
@@ -282,11 +274,112 @@ func (a *Accounts) SetCode(address flow.Address, code []byte) error {
+func (a *Accounts) GetStorageUsed(address flow.Address) (uint64, error) {
+	storageUsedRegister, err := a.getValue(address, false, keyStorageUsed)
+	if err != nil {
+		return 0, err
+	}
+
+	if len(storageUsedRegister) != uint64StorageSize {
+		return 0, fmt.Errorf("account %s storage used is not initialized or not initialized correctly", address.Hex())
+	}
+
+	storageUsed := binary.LittleEndian.Uint64(storageUsedRegister)
+	return storageUsed, nil
+}
+
+func (a *Accounts) setStorageUsed(address flow.Address, used uint64) error {
+	buffer := make([]byte, uint64StorageSize)
+	binary.LittleEndian.PutUint64(buffer, used)
+	return a.setValue(address, false, keyStorageUsed, buffer)
+}
+
+// GetValue returns a value stored in address' storage
+func (a *Accounts) GetValue(address flow.Address, key string) (flow.RegisterValue, error) {
+	return a.getValue(address, false, key)
+}
+
+func (a *Accounts) getValue(address flow.Address, isController bool, key string) (flow.RegisterValue, error) {
+	if isController {
+		return a.ledger.Get(string(address.Bytes()), string(address.Bytes()), key)
+	}
+	return a.ledger.Get(string(address.Bytes()), "", key)
+}
 
+// SetValue sets a value in address' storage
+func (a *Accounts) SetValue(address flow.Address, key string, value flow.RegisterValue) error {
+	return a.setValue(address, false, key, value)
+}
+
+func (a *Accounts) setValue(address flow.Address, isController bool, key string, value flow.RegisterValue) error {
+	err := a.updateRegisterSizeChange(address, isController, key, value)
+	if err != nil {
+		return fmt.Errorf("failed to update storage used by key %s on account %s: %w", key, address, err)
+	}
+
+	if isController {
+		a.ledger.Set(string(address.Bytes()), string(address.Bytes()), key, value)
+	} else {
+		a.ledger.Set(string(address.Bytes()), "", key, value)
+	}
 	return nil
 }
 
+func (a *Accounts) updateRegisterSizeChange(address flow.Address, isController bool, key string, value flow.RegisterValue) error {
+	if key == keyStorageUsed {
+		// size of this register is always uint64StorageSize
+		// don't double check this to save time and prevent recursion
+		return nil
+	}
+	oldValue, err := a.getValue(address, isController, key)
+	if err != nil {
+		return err
+	}
+
+	sizeChange := int64(len(value) - len(oldValue))
+	if sizeChange == 0 {
+		// register size has not changed. Nothing to do
+		return nil
+	}
+
+	oldSize, err := a.GetStorageUsed(address)
+	if err != nil {
+		return err
+	}
+
+	// two paths to avoid casting uint to int
+	var newSize uint64
+	if sizeChange < 0 {
+		absChange := uint64(-sizeChange)
+		if absChange > oldSize {
+			// should never happen
+			return fmt.Errorf("storage used by key %s on account %s would be negative", key, address.Hex())
+		}
+		newSize = oldSize - absChange
+	} else {
+		absChange := uint64(sizeChange)
+		newSize = oldSize + absChange
+	}
+
+	buffer := make([]byte, uint64StorageSize)
+	binary.LittleEndian.PutUint64(buffer, newSize)
```

Comment:
ideally, we keep this encoding part of the setStorageUsed()

Reply:
Good call, I agree 👍

Reply:
fixed!